### PR TITLE
Add 10 min timeout to `Test.Hspec.Extra.it`

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#2192 

# Overview

- [x] Add 10 min timeout to `it`.

# Comments

### CODE REVIEW

File: lib/test-utils/src/Test/Hspec/Extra.hs [hunk 1]
```diff
@@ -104,17 +113,38 @@ aroundAll acquire =
...
+                res2 <- try $ action ctx
+                case res2 of
+                    -- If the second try fails, return the first error. The
+                    -- second error might not be helpful.
+                    Left (e2 :: SomeException)
+                        | isAsyncCancelled e2 -> throwIO e1
+                        | otherwise           -> throwIO e1
+                    Right r2 -> return r2
+  where
+    isAsyncCancelled = isJust . fromException @AsyncCancelled
```
Comment (Anviking) [inline lib/test-utils/src/Test/Hspec/Extra.hs]: Without this, the `race` cancel is ignored and the timeout doesn't work. Good that I tested it.

**Comment by Anviking** [comment]
bors try

**Comment by Anviking** [comment]
bors try

File: lib/test-utils/src/Test/Hspec/Extra.hs [hunk 2]
```diff
@@ -103,18 +110,41 @@ aroundAll acquire =
...
+                | isAsyncCancelled e1 -> throwIO e1
+                | otherwise           -> do
+                res2 <- try $ action ctx
+                case res2 of
+                    -- If the second try fails, return the first error. The
+                    -- second error might not be helpful.
+                    Left (e2 :: SomeException)
+                        | isAsyncCancelled e2 -> throwIO e1
+                        | otherwise           -> throwIO e1
+                    Right r2 -> return r2
```
Comment (KtorZ) [inline lib/test-utils/src/Test/Hspec/Extra.hs]: :grimacing: Arg. Wouldn't it be better to separate both functions ? A `timeout` combinator would in principle not need to know anything about the execution flow of a particular spec. May I suggest a simpler variant:

```hs
it :: HasCallStack => String -> ActionWith ctx -> SpecWith ctx
it title action = specify title $ \ctx -> timeout _10min $ do
    action ctx
        `catch` (\(_ :: AsyncCancelled) -> return ())
        `catch` (\(e :: SomeException)  -> action ctx
        `catch` (\(_ :: SomeException)  -> throwIO e))
  where
    _10min = 600 :: NominalDiffTime

timeout :: NominalDiffTime -> IO () -> IO ()
timeout t action =
    race (threadDelay (micro t)) action >>= \case
        Right () ->
            return ()
        Left () ->
            expectationFailure $  "timed out in " <> show t
  where
    micro :: NominalDiffTime -> Int
    micro = round . (* 1000) . (* 1000)
```

Comment (Anviking) [reply lib/test-utils/src/Test/Hspec/Extra.hs]: 👍 nice.

**Review by KtorZ — APPROVED** [review]

**Comment by Anviking** [comment]
@KtorZ may you also look at #2193 too, which this depends on? 🙏

**Pull request merged**